tsdef/RemoteAudioTrackPublication.d.ts

```diff
@@ -0,0 +1,15 @@
+import { RemoteAudioTrack } from './RemoteAudioTrack';
+import { RemoteTrackPublication } from './RemoteTrackPublication';
+import { TwilioError } from './TwilioError';
+
+export class RemoteAudioTrackPublication extends RemoteTrackPublication {
+  kind: 'audio';
+  track: RemoteAudioTrack | null;
+
+  on(event: 'subscribed', listener: (track: RemoteAudioTrack) => any): this;
+  on(event: 'subscriptionFailed', listener: (error: TwilioError) => any): this;
+  on(event: 'trackDisabled', listener: () => any): this;
+  on(event: 'trackEnabled', listener: () => any): this;
+  on(event: 'unsubscribed', listener: (track: RemoteAudioTrack) => any): this;
+  on(event: string, listener: (...args: any[]) => any): this;
```

Comment:
Adding this line does get rid of the error: `Property 'on' in type 'RemoteAudioTrackPublication' is not assignable to the same property in base type 'RemoteTrackPublication'.`. It's a really simple approach, but in my eyes it has one drawback.

This approach works well when the developer spells the event name correctly. But when the event name is not spelled correctly, TS does not throw an error:

![image](https://user-images.githubusercontent.com/12685223/95511836-3b8f4200-0975-11eb-84be-9283cfc4a625.png)

It would be nice if TS could alert the user when an incorrect event name is used. Like this: 

![image](https://user-images.githubusercontent.com/12685223/95512145-a8a2d780-0975-11eb-9bb4-206b87c6ef51.png)

This is how it works with the diagnostics SDK (although an enum is used for the event name instead of a string). I think it would be great if we could implement something similar here too. I'm thinking about the best way to do this - it's a little tricky. I'd love to hear what others think.

Reply:
Yeah it would be nice but Diagnostics SDK can only do it because the inheritance is direct, i.e. we only inherit from `EventEmitter` once. TS is narrowing the type of `event` to the specific string, like `'trackSubscribed'`. Then that class will only accept `event: 'trackSubscribed'` which makes sense, but then when we inherit again then we can't define something like `event: 'trackAdded'` because `'trackAdded'` is not of type `'trackSubscribed'`.

Reply:
Yeah it's much easier with direct inheritance. I think I may have found a solution though:

This is the end result for the user. An object like `RemoteVideoTrackPublication` (which indirectly inherits from `EventEmitter`) lists all of its events and all of its inherited events:

![image](https://user-images.githubusercontent.com/12685223/95520728-2e7a4f00-0985-11eb-85a5-da2e187372f3.png)

It also knows the type of listener that's needed for a particular event:
![image](https://user-images.githubusercontent.com/12685223/95521113-0808e380-0986-11eb-88ff-99cee72ab7dd.png)


And it provides useful error messages when a non-existent event name is used:

![image](https://user-images.githubusercontent.com/12685223/95520783-4a7df080-0985-11eb-98e6-443d28b0c746.png)



To achieve this, we shouldn't use the node.js types for `EventEmitter` because it has no type-checking. Instead we should use something like this: https://github.com/binier/tiny-typed-emitter

This library provides type definitions for a node.js `EventEmitter` class that is type-aware. There's no JS in that library. It's a small library (and MIT licensed), so I just copied [this file](https://github.com/binier/tiny-typed-emitter/blob/master/lib/index.d.ts) into the twilio-video repo in order to get the following examples to work.

This is what TrackPublication.d.ts looks like:

```
import { ListenerSignature, TypedEventEmitter } from './TypedEventEmitter';
import { Track } from './Track';

export interface TrackPublicationEvents {
  trackDisabled: () => void;
  trackEnabled: () => void;
}

export class TrackPublication<T extends ListenerSignature<T>> extends TypedEventEmitter<TrackPublicationEvents & T> {
  isEnabled: boolean;
  kind: Track.Kind;
  trackName: string;
  trackSid: Track.SID;

  toJSON(): string;
  toString(): string;
}
```
A few things are happening here:
- The event names and their respective listeners are defined on the `TrackPublicationEvents` interface (I like how clean this is). 
- The `TrackPublication` class accepts a type variable `T` which extends `ListenerSignature<T>` 
- The `TrackPublication` class extends `TypedEventEmitter<TrackPublicationEvents & T>`. Note: here it is essential that the intersection type operator (`&`) is used to combine the `TrackPublicationEvents` interface with the type variable `T`.

This enables another class to extend TrackPublication and pass in it's own interface of event names and listeners as a type variable. Let's look at `RemoteTrackPublication.d.ts`:

```
import { RemoteTrack } from './types';
import { Track } from './Track';
import { TrackPublication } from './TrackPublication';
import { TwilioError } from './TwilioError';
import { ListenerSignature } from './TypedEventEmitter';

export interface RemoteTrackPublicationEvents {
  publishPriorityChanged: (priority: Track.Priority) => void;
  subscribed: (track: RemoteTrack) => void;
  subscriptionFailed: (error: TwilioError) => void;
  trackSwitchedOff: (track: RemoteTrack) => void;
  trackSwitchedOn: (track: RemoteTrack) => void;
  unsubscribed: (track: RemoteTrack) => void;
}

export class RemoteTrackPublication<T extends ListenerSignature<T>> extends TrackPublication<RemoteTrackPublicationEvents & T> {
  isSubscribed: boolean;
  publishPriority: Track.Priority;
  track: RemoteTrack | null;
}
```
Same thing here, but this class extends `TrackPublication` instead of `TypedEventEmitter`.  Again, the use of the type intersection operator here is important.

Finally, `RemoteVideoTrackPublication.d.ts`:
```
import { RemoteTrackPublication } from './RemoteTrackPublication';
import { RemoteVideoTrack } from './RemoteVideoTrack';
import { TwilioError } from './TwilioError';

export interface RemoteVideoTrackPublicationEvents {
  subscribed: (track: RemoteVideoTrack) => void;
  subscriptionFailed: (error: TwilioError) => void;
  trackSwitchedOff: (track: RemoteVideoTrack) => void;
  trackSwitchedOn: (track: RemoteVideoTrack) => void;
  unsubscribed: (track: RemoteVideoTrack) => void;
}

export class RemoteVideoTrackPublication extends RemoteTrackPublication<RemoteVideoTrackPublicationEvents> {
  kind: 'video';
  track: RemoteVideoTrack | null;
}
```

Since no other classes need to extend `RemoteVideoTrackPublication`, we don't need to set it up to receive a type variable.

What do you all think?

Reply:
I think that this is a really good idea. I would be more than happy to add this. I think this will be much more helpful to users in the future.